File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def decrease_sell_in(self, amount):
        self.sell_in -= amount

    def decrease_quality(self, amount):
        min_quality = 0
        self.quality -= amount
        if self.quality < min_quality:
            self.quality = min_quality
        

    def increase_quality(self, amount):
        max_quality = 50
        if self.quality < max_quality:
            self.quality += amount
            return
        self.quality = max_quality

    def is_past_sell_by_date(self):
        final_day = 0
        return self.sell_in < final_day

    def choose_and_convert_item(self, item):

        item_classes = {
            "Aged Brie": Brie,
            "Sulfuras, Hand of Ragnaros": Sulfaras,
            "Backstage passes to a TAFKAL80ETC concert": Ticket,
            "Conjured Mana Cake": Conjured,
            "Default": Default
        }
        chosen_class = item_classes.get(item.name, item_classes["Default"])

        return chosen_class(item.name, item.sell_in, item.quality)

    def process_all_items(self):
        for item in self.items:
            converted_item = self.choose_and_convert_item(item)
            converted_item.process_item()
            
            index_of_item = self.items.index(item)
            self.items[index_of_item] = converted_item

    def update_quality(self):
        self.process_all_items()


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

class Brie(Item, GildedRose):
    def __init__(self, name, sell_in, quality):
        super().__init__(name, sell_in, quality)

    def process_item(self):
        self.decrease_sell_in(1)

        if  self.is_past_sell_by_date():
            self.increase_quality(2)
        else:
            self.increase_quality(1)

class Ticket(Item, GildedRose):
    def __init__(self, name, sell_in, quality):
        super().__init__(name, sell_in, quality)

    def process_item(self):
        thresholds = {
            'first': 11,
            'last': 6
        }
        self.increase_quality(1)           

        if self.sell_in < thresholds['first']:
            self.increase_quality(1)

        if self.sell_in < thresholds['last']:
            self.increase_quality(1)

        self.decrease_sell_in(1)
        if self.is_past_sell_by_date():
            self.decrease_quality(self.quality)


class Sulfaras(Item, GildedRose):
    def __init__(self, name, sell_in, quality):
        super().__init__(name, sell_in, quality)

    def process_item(self):
        pass

class Conjured(Item, GildedRose):
    def __init__(self, name, sell_in, quality):
        super().__init__(name, sell_in, quality)

    def process_item(self):
        self.decrease_sell_in(1)
        self.decrease_quality(2)

class Default(Item, GildedRose):
    def __init__(self, name, sell_in, quality):
        super().__init__(name, sell_in, quality)

    def process_item(self):
        self.decrease_sell_in(1)
        
        if self.is_past_sell_by_date():
            self.decrease_quality(2)
        else:
            self.decrease_quality(1)

File: go/gildedrose/gildedrose.go
package gildedrose

type ItemVariant interface {
	ProcessItem()
}

type Item struct {
	Name            string
	SellIn, Quality int
}

func (i *Item) decreaseSellin(amount int) {
	i.SellIn -= amount
}

func (i *Item) decreaseQuality(amount int) {
	i.Quality -= amount
	if i.Quality < 0 {
		i.Quality = 0
	}
}

func (i *Item) increaseQuality(amount int) {
	if i.Quality < 50 {
		i.Quality += amount
		return
	}
	i.Quality = 50
}

func (i *Item) isPastSellByDate() bool {
	return i.SellIn < 0
}

// This function must be declared on the "parent" struct first
func (i *Item) ProcessItem() {
	// This handles the default behaviour
	i.decreaseSellin(1)

	if i.isPastSellByDate() {
		i.decreaseQuality(2)
	} else {
		i.decreaseQuality(1)
	}
}

type Brie struct {
	*Item
}

type Ticket struct {
	*Item
}

type Sulfuras struct {
	*Item
}

func (b *Brie) ProcessItem() {
	b.decreaseSellin(1)

	if b.isPastSellByDate() {
		b.increaseQuality(2)
	} else {
		b.increaseQuality(1)
	}

}

func (t *Ticket) ProcessItem() {

	type threshold struct {
		first int
		last  int
	}

	thresholds := threshold{
		first: 11,
		last:  6,
	}

	t.increaseQuality(1)

	if t.SellIn < thresholds.first {
		t.increaseQuality(1)
	}

	if t.SellIn < thresholds.last {
		t.increaseQuality(1)
	}

	t.decreaseSellin(1)
	if t.isPastSellByDate() {
		t.decreaseQuality(t.Quality)
	}

}

func (s *Sulfuras) ProcessItem() {

}

// Returning the interface will allow you to return multiple structs that share the interface
func ChooseAndCreateItem(item_variant *Item) ItemVariant {

	switch item_variant.Name {
	case "Aged Brie":
		return &Brie{item_variant}

	case "Sulfuras, Hand of Ragnaros":
		return &Sulfuras{item_variant}

	case "Backstage passes to a TAFKAL80ETC concert":
		return &Ticket{item_variant}

	default:
		return item_variant
	}
}

func RunEndOfDayForItems(items []*Item) {

	for _, item := range items {

		chosenStruct := ChooseAndCreateItem(item)
		chosenStruct.ProcessItem()
	}
}

func UpdateQuality(items []*Item) {

	RunEndOfDayForItems(items)

}
